### PR TITLE
Lens correction HIP and HOST support

### PR DESCRIPTION
**Pull request by snehaa8**

### CODE REVIEW

File: include/rppt_tensor_geometric_augmentations.h [hunk 1]
```diff
@@ -580,6 +580,111 @@ RppStatus rppt_flip_voxel_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDes
...
+ * \details Performs a remap operation using user specified remap tables for a batch of RGB(3 channel) / greyscale(1 channel) images with an NHWC/NCHW tensor layout. For each image, the output(x,y) = input(mapx(x, y), mapy(x, y)) for every (x,y) in the destination image.<br>
+ * - srcPtr depth ranges - Rpp8u (0 to 255), Rpp16f (0 to 1), Rpp32f (0 to 1), Rpp8s (-128 to 127).
+ * - dstPtr depth ranges - Will be same depth as srcPtr.
+ * \image html img150x150.png Sample Input
+ * \image html geometric_augmentations_remap_img150x150.png Sample Output
+ * \param [in] srcPtr source tensor in HOST memory
+ * \param [in] srcDescPtr source tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = 1/3)
+ * \param [out] dstPtr destination tensor in HOST memory
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param [in] rowRemapTable Rpp32u row numbers in HOST memory for every pixel in the input batch of images (1D tensor of size width * height * batchSize)
```
Comment (sampath1117) [inline include/rppt_tensor_geometric_augmentations.h:593]: Remap table is of type Rpp32f
So i feel it might be better to change the description from Rpp32u row numbers to Rpp32f row numbers

File: include/rppt_tensor_geometric_augmentations.h [hunk 2]
```diff
@@ -580,6 +580,111 @@ RppStatus rppt_flip_voxel_host(RppPtr_t srcPtr, RpptGenericDescPtr srcGenericDes
...
+ * \details Performs lens correction transforms on an image to compensate barrel lens distortion of RGB(3 channel) / greyscale(1 channel) images with an NHWC/NCHW tensor layout.<br>
+ * - srcPtr depth ranges - Rpp8u (0 to 255), Rpp16f (0 to 1), Rpp32f (0 to 1), Rpp8s (-128 to 127).
+ * - dstPtr depth ranges - Will be same depth as srcPtr.
+ * \image html lens_img640x480.png Sample Input
+ * \image html geometric_augmentations_lens_correction_img_640x480.png Sample Output
+ * \param [in] srcPtr source tensor in HOST memory
+ * \param [in] srcDescPtr source tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = 1/3)
+ * \param [out] dstPtr destination tensor in HOST memory
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param [in] rowRemapTable Rpp32u row numbers in HOST memory for every pixel in the input batch of images (1D tensor of size width * height * batchSize)
```
Comment (sampath1117) [inline include/rppt_tensor_geometric_augmentations.h:643]: Same comment
suggest to change description from Rpp32 to Rpp32f

File: src/include/cpu/rpp_cpu_common.hpp [hunk 3]
```diff
@@ -5033,8 +5033,8 @@ inline void compute_generic_bilinear_srclocs_1c_avx(__m256 &pSrcY, __m256 &pSrcX
 inline void compute_generic_bilinear_srclocs_3c_avx(__m256 &pSrcY, __m256 &pSrcX, RpptBilinearNbhoodLocsVecLen8 &srcLocs, __m256 *pBilinearCoeffs, __m256 &pSrcStrideH, __m256i *pxSrcStridesCHW, Rpp32s srcChannels, __m256 *pRoiLTRB, bool isSrcPKD3 = false)
 {
     __m256 pWeightParams[4], pSrcBilinearLTyx[4];
-    pSrcBilinearLTyx[0] = _mm256_floor_ps(pSrcY);                               // srcLT->y = (Rpp32s) srcY;
-    pSrcBilinearLTyx[1] = _mm256_floor_ps(pSrcX);                               // srcLT->x = (Rpp32s) srcX;
+    pSrcBilinearLTyx[0] = _mm256_floor_ps(pSrcY);     // srcLT->y = (Rpp32s) srcY;
```
Comment (sampath1117) [inline src/include/cpu/rpp_cpu_common.hpp:5036]: Please revert this change

File: src/include/cpu/rpp_cpu_common.hpp [hunk 4]
```diff
@@ -4995,10 +4995,10 @@ inline void compute_generic_bilinear_srclocs_and_interpolate(T *srcPtrChannel, R
 
     for (int c = 0; c < srcDescPtr->c; c++)
     {
-        dst[c] = (T)std::nearbyintf(((*(srcPtrChannel + srcLoc[0]) * bilinearCoeffs[0]) +        // TopRow R01 Pixel * coeff0
+        dst[c] = (T)((*(srcPtrChannel + srcLoc[0]) * bilinearCoeffs[0]) +       // TopRow R01 Pixel * coeff0
```
Comment (sampath1117) [inline src/include/cpu/rpp_cpu_common.hpp:4998]: Is this change necessary?
Would suggest to dump the old outputs and the current outputs with this change and see if those are matching

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 5]
```diff
@@ -3132,7 +3138,7 @@ inline void rpp_store8_f32pln1_to_i8pln1_avx(Rpp8s* dstPtr, __m256 &p)
 {
     __m256i px1 = _mm256_permute4x64_epi64(_mm256_packus_epi32(_mm256_cvtps_epi32(p), avx_px0), _MM_SHUFFLE(3,1,2,0));
     px1 = _mm256_sub_epi8(_mm256_packus_epi16(px1, avx_px0), avx_pxConvertI8);  /* Pack and add I8 conversion param */
-    _mm256_storeu_si256((__m256i *)(dstPtr), px1);                              /* store the 4 pixels in dst */
+    rpp_storeu_si64((__m128i *)(dstPtr), _mm256_castsi256_si128(px1));          /* store the 4 pixels in dst */
```
Comment (sampath1117) [inline src/include/cpu/rpp_cpu_simd.hpp:3141]: we are storing 24 values in output in this function before
but now we are storing only 64. Is this correct?

File: src/include/hip/rpp_hip_common.hpp [hunk 6]
```diff
@@ -55,12 +55,14 @@ typedef union { float f1[5];
...
 typedef union { float f1[8];    float2 f2[4];   float4 f4[2];                                   }   d_float8;
-typedef union { float f1[9];                                                                    }   d_float9;
+typedef union { float f1[9];    float3 f3[3];                                                   }   d_float9;
 typedef union { float f1[12];   float4 f4[3];                                                   }   d_float12;
+typedef union { float f1[14];                                                                   }   d_float14;
 typedef union { float f1[16];   float4 f4[4];   d_float8 f8[2];                                 }   d_float16;
 typedef union { float f1[24];   float2 f2[12];  float3 f3[8];   float4 f4[6];   d_float8 f8[3]; }   d_float24;
 
 // uint
+typedef union { uint ui1[4];    uint2 ui2[2];                                                   }   d_uint4;
```
Comment (sampath1117) [inline src/include/hip/rpp_hip_common.hpp:65]: There is already uint4 datatype available in HIP api
https://rocm.docs.amd.com/projects/HIP/en/latest/old/reference/kernel_language.html#short-vector-types

Please check and use it

File: src/include/hip/rpp_hip_common.hpp [hunk 7]
```diff
@@ -55,12 +55,14 @@ typedef union { float f1[5];
 typedef union { float f1[6];    float2 f2[3];                                                   }   d_float6;
 typedef union { float f1[7];                                                                    }   d_float7;
 typedef union { float f1[8];    float2 f2[4];   float4 f4[2];                                   }   d_float8;
-typedef union { float f1[9];                                                                    }   d_float9;
+typedef union { float f1[9];    float3 f3[3];                                                   }   d_float9;
 typedef union { float f1[12];   float4 f4[3];                                                   }   d_float12;
+typedef union { float f1[14];                                                                   }   d_float14;
```
Comment (sampath1117) [inline src/include/hip/rpp_hip_common.hpp:60]: Rajy has suggested in old gaussian filter PR to not add any new custom datatype anymore other than what 
is existing currently
@r-abishek Please share your thoughts here

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "rppdefs.h"
+#include "rpp_cpu_simd.hpp"
+#include "rpp_cpu_common.hpp"
+#include <omp.h>
+
+// Compute Inverse Matrix
```
Comment (sampath1117) [inline src/modules/cpu/kernel/lens_correction.hpp:30]: Would suggest to change comment as
`// Compute Inverse matrix (3x3)`

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    Rpp32u numThreads = handle.GetNumThreads();
+    omp_set_dynamic(0);
+#pragma omp parallel for num_threads(numThreads)
+    for(int batchCount = 0; batchCount < srcDescPtr->n; batchCount++)
+    {
+        Rpp32f *rowRemapTableTemp, *colRemapTableTemp;
+        rowRemapTableTemp = rowRemapTable + batchCount * tableDescPtr->strides.nStride;
+        colRemapTableTemp = colRemapTable + batchCount * tableDescPtr->strides.nStride;
+
+        Rpp32f *cameraMatrix = cameraMatrixTensor + batchCount * 9;
```
Comment (sampath1117) [inline src/modules/cpu/kernel/lens_correction.hpp:68]: Would suggest to add a comment mentioning like
cameraMatrix is a 3x3 matrix, hence * 9 is needed

File: src/modules/hip/kernel/lens_correction.hpp [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+__global__ void compute_remap_tables_hip_tensor(float *rowRemapTable,
+                                                float *colRemapTable,
+                                                d_float9 *cameraMatrixTensor,
+                                                d_float9 *inverseMatrixTensor,
+                                                d_float8 *distortionCoeffsTensor,
+                                                d_float9 *newCameraMatrixTensor,
+                                                uint2 remapTableStridesNH,
+                                                 RpptROIPtr roiTensorPtrSrc)
```
Comment (sampath1117) [inline src/modules/hip/kernel/lens_correction.hpp:33]: Please align params

File: src/modules/hip/kernel/lens_correction.hpp [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+    float *inverseMatrix = handle.GetInitHandle()->mem.mgpu.scratchBufferHip.floatmem;
+    hipLaunchKernelGGL(get_inverse_hip,
+                       dim3(1, 1, ceil((float)globalThreads_z/LOCAL_THREADS_Z)),
+                       dim3(1, 1, LOCAL_THREADS_Z),
+                       0,
+                       handle.GetStream(),
+                       (d_float9 *)cameraMatrix,
+                       (d_float9 *)inverseMatrix);
+    hipDeviceSynchronize();
```
Comment (sampath1117) [inline src/modules/hip/kernel/lens_correction.hpp:102]: Please remove the hipDeviceSynchronize() calls here
Since launching in same stream, this calls might not be needed

File: utilities/test_suite/HOST/Tensor_host.cpp [hunk 12]
```diff
@@ -1196,14 +1266,6 @@ int main(int argc, char **argv)
 
             if (reductionTypeCase)
             {
-                if(srcDescPtr->c == 3)
```
Comment (sampath1117) [inline utilities/test_suite/HOST/Tensor_host.cpp:1199]: why this is deleted?

**Review by sampath1117 — COMMENTED** [review]
Please address review comments

File: src/include/cpu/rpp_cpu_common.hpp (hunk 3 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp:5036]: Yes, reverted this in my latest commits, however this isn't getting reflected

File: src/include/cpu/rpp_cpu_common.hpp (hunk 4 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp:4998]: yep checked all QA

Comment (sampath1117) [reply src/include/cpu/rpp_cpu_common.hpp:4998]: Instead of checking QA
Would suggest to dump the outputs with and without this change

Since we have 1 pixel threshold for QA. Wanted to see if there is any difference made

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 5 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_simd.hpp:3141]: Only 1st quarter stores valid values, on storing register to size 256 it lead to output corruptions in remap kernel.

File: src/include/cpu/rpp_cpu_common.hpp (hunk 3 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp:5036]: It was added as part of "[https://github.com/r-abishek/rpp/pull/239/commits/1bcb8dc3e0c3120a82ed09778f9c3f2171ee4369](https://github.com/r-abishek/rpp/pull/239/commits/1bcb8dc3e0c3120a82ed09778f9c3f2171ee4369)"
https://github.com/r-abishek/rpp/blob/1bcb8dc3e0c3120a82ed09778f9c3f2171ee4369/src/include/cpu/rpp_cpu_common.hpp#L5006

File: src/include/hip/rpp_hip_common.hpp (hunk 6 above)
Comment (snehaa8) [reply src/include/hip/rpp_hip_common.hpp:65]: Sure

File: src/include/cpu/rpp_cpu_common.hpp (hunk 3 above)
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_common.hpp:5036]: It was just changed for compute_generic_bilinear_srclocs_1c_avx 
I commented for compute_generic_bilinear_srclocs_3c_avx

Please check the link to the code added in previous comment before replying

Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp:5036]: Oh yeah, fixed this.

File: src/modules/cpu/kernel/lens_correction.hpp (hunk 9 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/lens_correction.hpp:68]: Sure

File: src/include/cpu/rpp_cpu_common.hpp (hunk 4 above)
Comment (snehaa8) [reply src/include/cpu/rpp_cpu_common.hpp:4998]: We don't see any difference in outputs with this change, Previously we had added this to match HIP and HOST outputs right. Not very sure how come we don't see any effect of this change. Anyways I will revert my change.

File: utilities/test_suite/HOST/Tensor_host.cpp (hunk 12 above)
Comment (snehaa8) [reply utilities/test_suite/HOST/Tensor_host.cpp:1199]: This was a mistake during merging with latest branch, reverted this.

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+#include "rppdefs.h"
+#include "rpp_cpu_simd.hpp"
+#include "rpp_cpu_common.hpp"
+#include <omp.h>
+
+// Compute Inverse matrix (3x3)
+inline void get_inverse(float *mat, float *invMat)
+{
+    float det = mat[0] * (mat[4] * mat[8] - mat[7] * mat[5]) - mat[1] * (mat[3] * mat[8] - mat[5] * mat[6]) + mat[2] * (mat[3] * mat[7] - mat[4] * mat[6]);
+    if(det != 0)
```
Comment (r-abishek) [inline src/modules/cpu/kernel/lens_correction.hpp:34]: And if det = 0 what's our return?

File: src/modules/hip/kernel/lens_correction.hpp [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                       handle.GetStream(),
+                       rowRemapTable,
+                       colRemapTable,
+                       (d_float9 *)cameraMatrix,
+                       (d_float9 *)inverseMatrix,
+                       (d_float8 *)distanceCoeffs,
+                       (d_float9 *)newCameraMatrix,
+                       make_uint2(remapTableDescPtr->strides.nStride, remapTableDescPtr->strides.hStride),
+                       roiTensorPtrSrc);
+    hipDeviceSynchronize();
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:116]: Please remove deviceSynchronize. Stream synchronizes can be added between and after kernels

File: src/modules/hip/kernel/lens_correction.hpp [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    d_float9 *mat = &matTensor[id_z];
+    d_float9 *invMat = &invMatTensor[id_z];
+    float det = mat->f1[0] * (mat->f1[4] * mat->f1[8] - mat->f1[7] * mat->f1[5]) - mat->f1[1] * (mat->f1[3] * mat->f1[8] - mat->f1[5] * mat->f1[6]) + mat->f1[2] * (mat->f1[3] * mat->f1[7] - mat->f1[4] * mat->f1[6]);
+    if(det != 0)
+    {
+        float invDet = 1 / det;
+        invMat->f1[0] = (mat->f1[4] * mat->f1[8] - mat->f1[7] * mat->f1[5]) * invDet;
+        invMat->f1[1] = (mat->f1[2] * mat->f1[7] - mat->f1[1] * mat->f1[8]) * invDet;
+        invMat->f1[2] = (mat->f1[1] * mat->f1[5] - mat->f1[2] * mat->f1[4]) * invDet;
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:16]: This is long form determinant. Are their faster approaches, pls check

File: src/modules/hip/kernel/lens_correction.hpp [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+    float *rowRemapTableTemp = rowRemapTable + id_z * remapTableStridesNH.x + id_y * remapTableStridesNH.y + id_x;
+    float *colRemapTableTemp = colRemapTable + id_z * remapTableStridesNH.x + id_y * remapTableStridesNH.y + id_x;
+    float xCamera = id_y * inverseMatrix.f1[1] + inverseMatrix.f1[2] + id_x * inverseMatrix.f1[0];
+    float yCamera = id_y * inverseMatrix.f1[4] + inverseMatrix.f1[5] + id_x * inverseMatrix.f1[3];
+    float zCamera = id_y * inverseMatrix.f1[7] + inverseMatrix.f1[8] + id_x * inverseMatrix.f1[6];
+    float z = 1./zCamera, x = xCamera * z, y = yCamera * z;
+    float xSquare = x * x, ySquare = y * y;
+    float r2 = xSquare + ySquare, xyMul2 = 2 * x * y;
+    float kr = (1 + ((rCoeff[2] * r2 + rCoeff[1]) * r2 + rCoeff[0]) * r2) / (1 + ((rCoeff[5] * r2 + rCoeff[4]) * r2 + rCoeff[3]) *r2);
+    float colLoc = fx * (x * kr + tCoeff[0] * xyMul2 + tCoeff[1] * (r2 + 2 * xSquare)) + u0;
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:68]: Why can't this kernel be vectorized along x? All these arithmetic computations could be float4 computes too

File: src/modules/hip/kernel/lens_correction.hpp [hunk 17]
```diff
@@ -0,0 +1,119 @@
...
+__global__ void compute_remap_tables_hip_tensor(float *rowRemapTable,
+                                                float *colRemapTable,
+                                                d_float9 *cameraMatrixTensor,
+                                                d_float9 *inverseMatrixTensor,
+                                                d_float8 *distortionCoeffsTensor,
+                                                d_float9 *newCameraMatrixTensor,
+                                                uint2 remapTableStridesNH,
+                                                RpptROIPtr roiTensorPtrSrc)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:35]: Please vectorize this

**Review by r-abishek — CHANGES_REQUESTED** [review]

File: src/modules/hip/kernel/lens_correction.hpp (hunk 15 above)
Comment (sampath1117) [reply src/modules/hip/kernel/lens_correction.hpp:16]: @r-abishek 
we are computing here the inverse matrix values

The determinant  is computed in single line here
https://github.com/snehaa8/rpp/blob/sn/lens_correction/src/modules/cpu/kernel/lens_correction.hpp#L33

I have checked online and this is the fastest approach i have seen for computing inverse of a 3x3 matrix

File: src/modules/cpu/kernel/lens_correction.hpp (hunk 13 above)
Comment (sampath1117) [reply src/modules/cpu/kernel/lens_correction.hpp:34]: Currently leaving the code as is
The error code appraoch will have issues if we return an error code since we are processing batchwise
I have tried with identity matrix to replace if determinant is 0, but it is giving black images

**Comment by r-abishek** [comment]
> @r-abishek I have vectorized the HIP code and addressed all comments given except the case for handling if det = 0 Kindly note

@sampath1117 Lets go over this on a call

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+                pXCamera = _mm256_add_ps(pXCamera, pXCameraIncrement);
+                pYCamera = _mm256_add_ps(pYCamera, pYCameraIncrement);
+                pZCamera = _mm256_add_ps(pZCamera, pZCameraIncrement);
+            }
+            for(; vectorLoopCount < width; vectorLoopCount++)
+            {
+                Rpp32f z = 1./zCamera, x = xCamera * z, y = yCamera * z;
+                Rpp32f xSquare = x * x, ySquare = y * y, r2 = xSquare + ySquare;
+                Rpp32f xyMul2 = 2 * x * y;
+                Rpp32f r4 = r2 * r2;
```
Comment (r-abishek) [inline src/modules/cpu/kernel/lens_correction.hpp:172]: Where is this r4 used?

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+                __m256 pFac1 = _mm256_mul_ps(pX, pKR);
+                __m256 pFac2 = _mm256_mul_ps(pTCoeff[0], p2xy);
+                __m256 pFac3 = _mm256_mul_ps(pTCoeff[1], _mm256_fmadd_ps(avx_p2, pX2, pR2));
+                __m256 pSrcCol = _mm256_fmadd_ps(pFx, _mm256_add_ps(pFac1, _mm256_add_ps(pFac2, pFac3)), pU0);
+
+                // float rowLoc = fy * (y * kr + tCoeff[0] * (r2 + 2 * ySquare) + tCoeff[1] * xyMul2) + v0;
+                pFac1 = _mm256_mul_ps(pY, pKR);
+                pFac2 = _mm256_mul_ps(pTCoeff[1], p2xy);
+                pFac3 = _mm256_mul_ps(pTCoeff[0], _mm256_fmadd_ps(avx_p2, pY2, pR2));
+                __m256 pSrcRow = _mm256_fmadd_ps(pFy, _mm256_add_ps(pFac1, _mm256_add_ps(pFac2, pFac3)), pV0);
```
Comment (r-abishek) [inline src/modules/cpu/kernel/lens_correction.hpp:155]: If pFac2 + pFac3 is being done in the end anyway, pFac2 could directly be _mm256_fmadd_ps(<>) on L153?

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+            }
+            for(; vectorLoopCount < width; vectorLoopCount++)
+            {
+                Rpp32f z = 1./zCamera, x = xCamera * z, y = yCamera * z;
+                Rpp32f xSquare = x * x, ySquare = y * y, r2 = xSquare + ySquare;
+                Rpp32f xyMul2 = 2 * x * y;
+                Rpp32f r4 = r2 * r2;
+                Rpp32f kr = (1 + ((rCoeff[2] * r2 + rCoeff[1]) * r2 + rCoeff[0]) * r2) / (1 + ((rCoeff[5] * r2 + rCoeff[4]) * r2 + rCoeff[3]) *r2);
+                Rpp32f colLoc = fx * (x * kr + tCoeff[0] *xyMul2 + tCoeff[1] * (r2 + 2 * xSquare)) + u0;
+                Rpp32f rowLoc = fy * (y * kr + tCoeff[0] * (r2 + 2 * ySquare ) + tCoeff[1] *xyMul2) + v0;
```
Comment (r-abishek) [inline src/modules/cpu/kernel/lens_correction.hpp:175]: Please add parentheses for all these, fix 4-5 space-issues in L174 after *, L175 before ) etc.
Also, I think the L174 for example can be done in 4 fma + 1 mul instruction. But above I see 3 mul + 2 fma + 2 add used.
Pls check correspondingly for L175 too.

File: src/modules/hip/kernel/lens_correction.hpp [hunk 21]
```diff
@@ -0,0 +1,159 @@
...
+                       0,
+                       handle.GetStream(),
+                       rowRemapTable,
+                       colRemapTable,
+                       reinterpret_cast<d_float9 *>(cameraMatrix),
+                       reinterpret_cast<d_float9 *>(inverseMatrix),
+                       reinterpret_cast<d_float8 *>(distanceCoeffs),
+                       make_uint2(remapTableDescPtr->strides.nStride, remapTableDescPtr->strides.hStride),
+                       roiTensorPtrSrc);
+    hipStreamSynchronize(handle.GetStream());
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:156]: Lets remove this considering they are in the same stream

**Comment by r-abishek** [comment]
@sampath1117 Please add fixes for minor comments above

File: src/modules/cpu/kernel/lens_correction.hpp (hunk 18 above)
Comment (sampath1117) [reply src/modules/cpu/kernel/lens_correction.hpp:172]: Removed it
Should have been added by mistake

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+                pZCamera = _mm256_add_ps(pZCamera, pZCameraIncrement);
+            }
+            for(; vectorLoopCount < width; vectorLoopCount++)
+            {
+                Rpp32f z = 1./zCamera, x = xCamera * z, y = yCamera * z;
+                Rpp32f xSquare = x * x, ySquare = y * y, r2 = xSquare + ySquare;
+                Rpp32f xyMul2 = 2 * x * y;
+                Rpp32f kr = (1 + ((((rCoeff[2] * r2) + rCoeff[1]) * r2) + rCoeff[0]) * r2) / (1 + ((((rCoeff[5] * r2) + rCoeff[4]) * r2) + rCoeff[3]) * r2);
+                Rpp32f colLoc = std::fmaf(fx, (std::fmaf(tCoeff[1], (std::fmaf(2, xSquare, r2)), std::fmaf(x, kr, (tCoeff[0] * xyMul2)))), u0);
+                Rpp32f rowLoc = std::fmaf(fy, (std::fmaf(tCoeff[0], (std::fmaf(2, ySquare, r2)), std::fmaf(y, kr, (tCoeff[1] * xyMul2)))), v0);
```
Comment (r-abishek) [inline src/modules/cpu/kernel/lens_correction.hpp:173]: The rowLoc and colLoc ops look pretty uniform in raw C with 4 fmafs and 1 multiply above.
But the AVX side looks different - could we match to the same 4fmafs and 1 multiply? And secondly, just copy the raw C code line as a comment above the AVX code so its really clear what equivalent operation the AVX is doing.

File: src/modules/cpu/kernel/lens_correction.hpp [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+                pXCamera = _mm256_add_ps(pXCamera, pXCameraIncrement);
+                pYCamera = _mm256_add_ps(pYCamera, pYCameraIncrement);
+                pZCamera = _mm256_add_ps(pZCamera, pZCameraIncrement);
+            }
+            for(; vectorLoopCount < width; vectorLoopCount++)
+            {
+                Rpp32f z = 1./zCamera, x = xCamera * z, y = yCamera * z;
+                Rpp32f xSquare = x * x, ySquare = y * y, r2 = xSquare + ySquare;
+                Rpp32f xyMul2 = 2 * x * y;
+                Rpp32f kr = (1 + ((((rCoeff[2] * r2) + rCoeff[1]) * r2) + rCoeff[0]) * r2) / (1 + ((((rCoeff[5] * r2) + rCoeff[4]) * r2) + rCoeff[3]) * r2);
```
Comment (r-abishek) [inline src/modules/cpu/kernel/lens_correction.hpp:171]: In this case since the AVX side shows 3 fmaf for numerator and 3 fmaf for denominator, can we use the same std::fmaf here?
Similarly copy the same raw C line you finalize on, as a comment above the AVX code - to make it easier for the reader.

**Review by r-abishek — CHANGES_REQUESTED** [review]
@sampath1117 Added two more comments.

File: src/modules/hip/kernel/lens_correction.hpp [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+    d_float8 r2_f8, kr_f8, kr1_f8, kr2_f8;
+    rpp_hip_math_add8(&xSquare_f8, &ySquare_f8, &r2_f8);                                        // float r2 = xSquare + ySquare
+
+    d_float8 r2Cube_f8, r2Square_f8;
+    rpp_hip_math_multiply8(&r2_f8, &r2_f8, &r2Square_f8);                                       // float r2Square = r2 * r2;
+    rpp_hip_math_multiply8(&r2Square_f8, &r2_f8, &r2Cube_f8);                                   // float r2Cube = r2Square * r2;
+
+    // float kr = (1 + (radialCoeff[2] * r2Cube) + (radialCoeff[1] * r2Square) + (radialCoeff[0]) * r2)) / (1 + (radialCoeff[5] * r2Cube) + (radialCoeff[4] * r2Square) + (radialCoeff[3]) *r2))
+    kr1_f8.f4[0] = (one_f4 + static_cast<float4>(radialCoeff[2]) * r2Cube_f8.f4[0] + static_cast<float4>(radialCoeff[1]) *  r2Square_f8.f4[0] + static_cast<float4>(radialCoeff[0]) *  r2_f8.f4[0]);
+    kr1_f8.f4[1] = (one_f4 + static_cast<float4>(radialCoeff[2]) * r2Cube_f8.f4[1] + static_cast<float4>(radialCoeff[1]) *  r2Square_f8.f4[1] + static_cast<float4>(radialCoeff[0]) *  r2_f8.f4[1]);
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:92]: To avoid this many static_cast between L90-L94, can't we assign a single d_float24 radialCoeff_f24 on L61 as 6 float4s?
Something like this will work?
```
d_float24 radialCoeff_f24 = {static_cast<float4>(distortionCoeffs_f8.f1[0]),
                                              static_cast<float4>(distortionCoeffs_f8.f1[1]),
                                              static_cast<float4>(distortionCoeffs_f8.f1[4]),
                                              static_cast<float4>(distortionCoeffs_f8.f1[5]),
                                              static_cast<float4>(distortionCoeffs_f8.f1[6]),
                                              static_cast<float4>(distortionCoeffs_f8.f1[7])};
```

File: src/modules/hip/kernel/lens_correction.hpp [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+    rpp_hip_math_multiply8(&x_f8, &y_f8, &xyMul2_f8);
+    rpp_hip_math_multiply8_const(&xyMul2_f8, &xyMul2_f8, two_f4);                               // float xyMul2 = 2 * x * y
+
+    d_float8 colLoc_f8, rowLoc_f8;
+    rpp_hip_math_multiply8_const(&xSquare_f8, &xSquare_f8, two_f4);                             // xSquare = xSquare * 2;
+    rpp_hip_math_multiply8_const(&ySquare_f8, &ySquare_f8, two_f4);                             // ySquare = ySquare * 2;
+
+    // float colLoc = cameraMatrix[0] * (x * kr + tangentialCoeff[0] * xyMul2 + tangentialCoeff[1] * (r2 + 2 * xSquare)) + cameraMatrix[2];
+    colLoc_f8.f4[0] = static_cast<float4>(cameraMatrix_f9.f1[0]) * ((x_f8.f4[0] * kr_f8.f4[0]) + (static_cast<float4>(tangentialCoeff[0]) * xyMul2_f8.f4[0]) + (static_cast<float4>(tangentialCoeff[1]) * (r2_f8.f4[0] + xSquare_f8.f4[0]))) + static_cast<float4>(cameraMatrix_f9.f1[2]);
+    colLoc_f8.f4[1] = static_cast<float4>(cameraMatrix_f9.f1[0]) * ((x_f8.f4[1] * kr_f8.f4[1]) + (static_cast<float4>(tangentialCoeff[0]) * xyMul2_f8.f4[1]) + (static_cast<float4>(tangentialCoeff[1]) * (r2_f8.f4[1] + xSquare_f8.f4[1]))) + static_cast<float4>(cameraMatrix_f9.f1[2]);
```
Comment (r-abishek) [inline src/modules/hip/kernel/lens_correction.hpp:107]: Similar comment for L106/107 and L110/111

**Review by r-abishek — APPROVED** [review]
